### PR TITLE
Add RateLimit config to serviceDefaults

### PR DESCRIPTION
**Pull request by kisunji**
Changes proposed in this PR:
- Adds the RateLimits field to `serviceDefaults` CRD to match the addition in https://github.com/hashicorp/consul/pull/18583

How I've tested this PR:
- Added unit tests

How I expect reviewers to test this PR:

Checklist:
- [x] Tests added
- [x] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml [hunk 1]
```diff
@@ -185,6 +185,63 @@ spec:
...
+                            requestsMaxBurst:
+                              type: integer
+                            requestsPerSecond:
+                              type: integer
+                          required:
+                          - pathExact
+                          - pathPrefix
+                          - pathRegex
+                          - requestsMaxBurst
+                          - requestsPerSecond
```
Comment (kisunji) [inline control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml:235]: I'm not sure how this was generated and what I need to do about this. Do I edit the "required" fields directly?

**Comment by lkysow** [comment]
Just a driveby but I think we used to also try and exercise new fields in our acceptance tests?

**Comment by wilkermichael** [comment]
Another drive by, haven't had a chance to fully review. But did you try creating the resource locally as part of your changes? Just to verify it exists properly in Consul?

**Comment by kisunji** [comment]
> Another drive by, haven't had a chance to fully review. But did you try creating the resource locally as part of your changes? Just to verify it exists properly in Consul?

It's not a new resource but a field added to service-defaults. In an acceptance test I added an assertion that the field exists with the right value from the fixture yaml.

I'll spend time writing a proper acceptance test eventually. I need to shift focus to some multiport/v2 work

File: control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml [hunk 2]
```diff
@@ -185,6 +185,69 @@ spec:
                   unlock usage of the service-splitter and service-router config entries
                   for a service.
                 type: string
+              rateLimits:
+                description: RateLimits is rate limiting configuration that is applied
+                  to inbound traffic for a service. Rate limiting is a Consul enterprise
```
Comment (wilkermichael) [inline control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml:190]:
```suggestion
                  to inbound traffic for a service. Rate limiting is a Consul Enterprise
```

File: control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml [hunk 3]
```diff
@@ -185,6 +185,69 @@ spec:
...
+                properties:
+                  instanceLevel:
+                    description: InstanceLevel represents rate limit configuration
+                      that is applied per service instance.
+                    properties:
+                      requestsMaxBurst:
+                        description: "RequestsMaxBurst is the maximum number of requests
+                          that can be sent in a burst. Should be equal to or greater
+                          than RequestsPerSecond. If unset, defaults to RequestsPerSecond.
+                          \n Internally, this is the maximum size of the token bucket
```
Comment (wilkermichael) [inline control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml:201]:
```suggestion
                          Internally, this is the maximum size of the token bucket
```

File: control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml [hunk 4]
```diff
@@ -185,6 +185,69 @@ spec:
...
+                          used for rate limiting."
+                        type: integer
+                      requestsPerSecond:
+                        description: "RequestsPerSecond is the average number of requests
+                          per second that can be made without being throttled. This
+                          field is required if RequestsMaxBurst is set. The allowed
+                          number of requests may exceed RequestsPerSecond up to the
+                          value specified in RequestsMaxBurst. \n Internally, this
```
Comment (wilkermichael) [inline control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml:209]:
```suggestion
                          value specified in RequestsMaxBurst. Internally, this
```

File: control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml [hunk 5]
```diff
@@ -185,6 +185,69 @@ spec:
...
+                description: RateLimits is rate limiting configuration that is applied
+                  to inbound traffic for a service. Rate limiting is a Consul enterprise
+                  feature.
+                properties:
+                  instanceLevel:
+                    description: InstanceLevel represents rate limit configuration
+                      that is applied per service instance.
+                    properties:
+                      requestsMaxBurst:
+                        description: "RequestsMaxBurst is the maximum number of requests
```
Comment (wilkermichael) [inline control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml:198]: For consistency remove quotes, do this elsewhere as well

File: control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml [hunk 6]
```diff
@@ -185,6 +185,69 @@ spec:
...
+                          field is required if RequestsMaxBurst is set. The allowed
+                          number of requests may exceed RequestsPerSecond up to the
+                          value specified in RequestsMaxBurst. \n Internally, this
+                          is the refill rate of the token bucket used for rate limiting."
+                        type: integer
+                      routes:
+                        description: Routes is a list of rate limits applied to specific
+                          routes. Overrides any top-level configuration.
+                        items:
+                          description: InstanceLevelRouteRateLimits represents rate
```
Comment (wilkermichael) [inline control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml:216]: Should this description be routes?

File: control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml [hunk 7]
```diff
@@ -185,6 +185,69 @@ spec:
...
+                            limit configuration applied to a route matching one of
+                            PathExact/PathPrefix/PathRegex.
+                          properties:
+                            pathExact:
+                              type: string
```
Comment (wilkermichael) [inline control-plane/config/crd/bases/consul.hashicorp.com_servicedefaults.yaml:221]: missing a description?

File: charts/consul/templates/crd-servicedefaults.yaml [hunk 8]
```diff
@@ -192,6 +192,72 @@ spec:
                   unlock usage of the service-splitter and service-router config entries
                   for a service.
                 type: string
+              rateLimits:
```
Comment (wilkermichael) [inline charts/consul/templates/crd-servicedefaults.yaml:195]: Same comments as `config/crd/bases` I forget which one is auto-generated

File: .changelog/2844.txt [hunk 9]
```diff
@@ -0,0 +1,3 @@
+```release-note:feature
+mesh: **(Enterprise only)** Adds rate limiting config to serviceDefaults CRD
```
Comment (wilkermichael) [inline .changelog/2844.txt:2]:
```suggestion
helm: (Consul Enterprise) Adds rate limiting config to serviceDefaults CRD
```

Comment (wilkermichael) [reply .changelog/2844.txt:2]: Just to be consistent with other changlog entries

File: acceptance/tests/fixtures/bases/crds-oss/servicedefaults.yaml [hunk 10]
```diff
@@ -31,6 +31,16 @@ spec:
...
   balanceInboundConnections: "exact_balance"
+  rateLimits:
+    instanceLevel:
+      requestsPerSecond: 1234
+      requestsMaxBurst: 2345
+      routes:
+        - pathExact: "/foo"
+          requestsPerSecond: 222
+          requestsMaxBurst: 333
+        - pathPrefix: "/admin"
```
Comment (wilkermichael) [inline acceptance/tests/fixtures/bases/crds-oss/servicedefaults.yaml:42]: add one with path regex?

File: control-plane/api/v1alpha1/servicedefaults_types.go [hunk 11]
```diff
@@ -214,6 +218,152 @@ type ServiceDefaultsDestination struct {
...
+	}
+
+	if irl.RequestsPerSecond == 0 && irl.RequestsMaxBurst > 0 {
+		allErrs = append(allErrs,
+			field.Invalid(path.Child("requestsPerSecond"),
+				irl.RequestsPerSecond,
+				"RequestsPerSecond must be greater than 0 if RequestsMaxBurst is set"))
+	}
+
+	if irl.RequestsMaxBurst < 0 {
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/servicedefaults_types.go:304]: error says "RequestsMaxBurst must be greater than 0" should this be `<=0` in the conditional in that case?

File: control-plane/api/v1alpha1/servicedefaults_types.go [hunk 12]
```diff
@@ -214,6 +218,152 @@ type ServiceDefaultsDestination struct {
...
+
+	for i, route := range irl.Routes {
+		if exact, prefix, regex := route.PathExact != "", route.PathPrefix != "", route.PathRegex != ""; (!exact && !prefix && !regex) ||
+			(exact && prefix) || (exact && regex) || (prefix && regex) {
+			allErrs = append(allErrs, field.Required(
+				path.Child("routes").Index(i),
+				"Route must define exactly one of PathExact, PathPrefix, or PathRegex"))
+		}
+
+		isRatelimitSet = isRatelimitSet || route.RequestsPerSecond > 0
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/servicedefaults_types.go:319]: Can you add a comment here for this check. Just want to make sure, we only require one route to have `route.RequestPerSecond` set for isRatelimitSet to be set

Comment (wilkermichael) [reply control-plane/api/v1alpha1/servicedefaults_types.go:319]: nit: 
```suggestion
		isRateLimitSet = isRateLimitSet || route.RequestsPerSecond > 0
```

File: control-plane/api/v1alpha1/servicedefaults_types.go [hunk 13]
```diff
@@ -214,6 +218,152 @@ type ServiceDefaultsDestination struct {
...
+
+		// Unlike top-level RequestsPerSecond, any route MUST have a RequestsPerSecond defined.
+		if route.RequestsPerSecond <= 0 {
+			allErrs = append(allErrs, field.Invalid(
+				path.Child("routes").Index(i).Child("requestsPerSecond"),
+				route.RequestsPerSecond, "RequestsPerSecond must be greater than 0"))
+		}
+
+		if route.RequestsMaxBurst < 0 {
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/servicedefaults_types.go:328]: should this be `<= 0` sine the error says `must be greater than 0`?

Comment (wilkermichael) [reply control-plane/api/v1alpha1/servicedefaults_types.go:328]: if that's the case should try and test these boundaries in the unit test as well

**Review by wilkermichael — COMMENTED** [review]
Great work on this @kisunji, a few minor comments. I also just want to double check some of the verification logic around including `0` as a valid value.

**Comment by wilkermichael** [comment]
> > Another drive by, haven't had a chance to fully review. But did you try creating the resource locally as part of your changes? Just to verify it exists properly in Consul?
> 
> It's not a new resource but a field added to service-defaults. In an acceptance test I added an assertion that the field exists with the right value from the fixture yaml.
> 
> I'll spend time writing a proper acceptance test eventually. I need to shift focus to some multiport/v2 work

Did you test that locally? The acceptance tests in the pipeline are currently not working, but they should work locally if you build dataplane yourself.

File: control-plane/api/v1alpha1/servicedefaults_types.go (hunk 11 above)
Comment (kisunji) [reply control-plane/api/v1alpha1/servicedefaults_types.go:304]: I'll align it with RequestsPerSecond and say "positive" instead. 0 / unset is a valid value.

File: control-plane/api/v1alpha1/servicedefaults_types.go (hunk 12 above)
Comment (kisunji) [reply control-plane/api/v1alpha1/servicedefaults_types.go:319]: The validation below ensures that every route has a >0 RequestsPerSecond. But I'll add a comment to clarify

**Review by wilkermichael — APPROVED** [review]
Thanks for addressing my comments!

File: control-plane/api/v1alpha1/servicedefaults_types.go (hunk 11 above)
Comment (kisunji) [reply control-plane/api/v1alpha1/servicedefaults_types.go:304]: Oh yes you're right!

Comment (wilkermichael) [reply control-plane/api/v1alpha1/servicedefaults_types.go:304]: nevermind, all good, disregard I read that wrong

Comment (kisunji) [reply control-plane/api/v1alpha1/servicedefaults_types.go:304]: 😆

**Pull request merged**